crates/node-core/src/init.rs

```diff
@@ -379,10 +379,40 @@ pub fn init_from_state_dump<DB: Database>(
         line.clear();
     }
 
+    // compute and compare state root
+    //
+    // note: we compute it even if we have nothing to compare to because we set the stage
+    // checkpoints further up.
+    let computed_state_root = compute_state_root(&provider_rw)?;
+    if let Some(expected_state_root) = expected_state_root {
+        if computed_state_root != expected_state_root {
+            error!(target: "reth::cli", ?computed_state_root, ?expected_state_root, "Computed state root does not match state root in state dump");
+            eyre::bail!("Computed state root differs from state root in state dump. Got {:?}, expected {:?}", computed_state_root, expected_state_root);
+        }
+    } else {
+        warn!(target: "reth::cli", "No state root found in file, skipping verification.");
+    }
+    provider_rw.commit()?;
+
     Ok(hash)
 }
 
-#[derive(Debug, Serialize, Deserialize)]
+/// Computes the state root (from scratch) based on the accounts and storages present in the
+/// database.
+fn compute_state_root<DB: Database>(provider: &DatabaseProviderRW<DB>) -> eyre::Result<StateRoot> {
+    trace!(target: "reth::cli", "Computing state root");
+
+    let tx = provider.tx_ref();
+    let (root, updates) = StateRootComputer::from_tx(tx)
+        .with_intermediate_state(None)
+        .with_no_threshold()
+        .root_with_updates()?;
```

Comment:
nit: `root_with_updates` already sets `with_no_threshold` and `with_intermediate_state(None)` is a noop
```suggestion
        .root_with_updates()?;
```